### PR TITLE
Ensure pool processes have closed

### PR DESCRIPTION
**Pull request by joelN123**
In the main training file, make sure pool processes have closed down after training has finished by manually shutting them down.

## Motivation and Context
This is to solve the bug described in issue https://github.com/blue-oil/blueoil/issues/725 where the training loop has finished but the program doesn't end. It is especially a concern when using time-costed services, if the program still runs after training has finished, it is a waste of costly resource.

## Description
A simple function is added, which accepts a dataset object, and terminates, closes and joins pools from the prefetcher of that dataset. This is only relevant if prefetch is being used, so first it checks if enable_prefetch is true.

The function is called at the end of training on each of the datasets: train, validation, and train_validation_saving (if there is one).

Honestly, I'm not sure why this fixes the problem, the dataset prefetcher should shut down own its own without it. But it doesn't in certain important cases.

## How has this been tested?
How to reproduce the bug is explained in issue https://github.com/blue-oil/blueoil/issues/725

To test that this fix does work, I ran training under the same settings, except with the code change that is in this pull request, and it runs fine, and fixes the bug.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: lmnet/executor/train.py [hunk 1]
```diff
@@ -58,6 +58,13 @@ def setup_dataset(config, subset, rank):
     return DatasetIterator(dataset, seed=rank, enable_prefetch=enable_prefetch)
 
 
+def close_dataset(dataset):
```
Comment (iizukak) [inline lmnet/executor/train.py]: @joelN123 
Hi, thank you for try to fix the bug!!
`train.py` is the main loop of training.
It's better to implement `close_dataset` feature in `DatasetItretaor` class.
How about that?

Comment (joelN123) [reply lmnet/executor/train.py]: ah, right, as a method of datasetiterator is better, yes.

Comment (joelN123) [reply lmnet/executor/train.py]: I will make a change now and test it still is working. Thank you for the advice

Comment (joelN123) [reply lmnet/executor/train.py]: oh, I should also remove the == True in the if statement

Comment (joelN123) [reply lmnet/executor/train.py]: I made that change, and it runs fine (hangs up properly) still.

File: lmnet/lmnet/datasets/dataset_iterator.py [hunk 2]
```diff
@@ -306,6 +306,12 @@ def get_shuffle_index(self):
         self.seed += 1
         return random_indices
 
+    def close_dataset(self):
```
Comment (tkng) [inline lmnet/lmnet/datasets/dataset_iterator.py]: Since what is closed here is trivial, how about changing the method name to `close`?

Comment (joelN123) [reply lmnet/lmnet/datasets/dataset_iterator.py]: yep, that's better name, thank you

**Comment by joelN123** [comment]
run test

**Comment by joelN123** [comment]
the "looks good to me" is also needed as a separate review?

**Comment by Joeper214** [comment]
> the "looks good to me" is also needed as a separate review?

I think It's not required. This PR can be merged of you have OA and RA.

**Pull request merged**